### DISCUSSION TITLE
Regarding "code tidy and improvement"

### DISCUSSION THREAD

**Discussion by qrnch-jan** [avastmick/google-generative-ai-api-rs-#40] [Ideas]
I've been experimenting with this crate a little and I tried to do some relatively simple operations, but failed at it at first.  This was both due to my own ignorance (of how the generative AI's, and models, work and how the REST API work on google's part), but there are some things the crate API could have done to help steer me in the right direction.

Specifically; I wanted to insert a "system instruction" for a query, and found that I could do so using the `beta` feature.  However, when the program compiled and ran, I would get a bad request back without any further information.  I eventually discovered that this was due to not using an appropriate model.

While this was obviously my fault for not knowing the default model doesn't support system instructions, I feel that this is the sort of thing Rust's type system could have helped me realize:  The compiler could have stopped me from trying to insert a system instruction with the default model.

I will preface this by emphasizing that I am very new to all these things, so what I'm suggesting might not be possible -- but I'd like to do some brainstorming if it is.

I'd like there to be a way to construct requests using builder patterns that are strongly typed.  You start by spawning a builder from a specific model specifier, so that all the builders that follow will only allow model-allowed sub-builders.

```
Request::builder(Model::Something)
  .system_instruction(SystemInstructionBuilder::new()
    .add("You are Sadie Adler.")
    .add("You are in a good mood.")
    .build()
  )
  .content(ContentsBuilder::new()
    .user(PartBuilder::new()
      .text("This is some text")
      .build()
    )
    .build()
  )
  .build()
```

I'd like to start experimenting with such a design in a branch and see if it's worth pursuing further.  However, before doing so I want to know if this is something of interest to the project.

Just to be clear:  The idea is not to replace existing structures, but rather build a more typesafe Builder-pattern wrapper around it.

**Comment by avastmick** [comment]
Thanks, @qrnch-jan. I agree with pretty much everything you say—trust me, I am very aware that usage is not ideal and that the more help the user gets to build great things, the better! Please feel free to try making the issues you detail easier for the user to resolve.

I want to keep the library as simple as possible and, where possible, as small as possible. While I understand the steep initial developer ramp-up, I do not want the library to seek to improve Google's documentation, make it more complex to maintain, or grow in compiled size.

I have always struggled with the Google documentation. It contains errors, inconsistencies, and misleading or vague aspects that have been hard to address—hence why I documented the Google usage docs directly in the doc comments. The `bad request` `400` "feature" is painful and the Google go-to error. For example, if your API key has been revoked, you get a `400` rather than an expected `403` or `401`, according to the [RFC](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html). When I first wrote the library, there were even undocumented API structural variations between the public API and the Vertex equivalent. This is improving, but tracking changes is non-trivial.

I modelled the crate (copied!?) from [openai-api-rs](https://github.com/dongri/openai-api-rs), which has a great community of devs who have created a stable, well-enough documented library that is easy to use. I recommend having a look. They seem to have a good balance on documentation, usage and maintainability.

I guess the only thing I would be cautious about is complicating the library. When I started, one of my goals was to work towards parity with the [generative-ai-python](https://github.com/google/generative-ai-python) library. It is good to check on that library to see how they track the Google features and documentation - there are some interesting threads on design choices and how to manage complexity. Honestly, I now think this library should only focus on the Gemini models and be as simple as possible.

I look forward to further contributions from you :pray:

Many thanks, Mick.

**Reply by qrnch-jan** [reply]
I can fully get behind all those design choices; being inspired by a reference implementation is generally a good idea.  I believe that what I'm looking to do is fully compatible with that goal.   I've been thinking about it some more, and there may even be a zero-cost way to get most of the type-safety I'm looking for.

I'm gonna think some more on this and return when I have more more concrete ideas.
